File: C++/Subsets.cpp
#include<iostream>
#include<vector>
#include<algorithm>

static bool cmp(int x,int y){
    return x<y;
}


void generate(vector<vector<int> >&vec,vector<int> &st,vector<int>& S,int pos,int len)
    {
        if(pos==len)
            {
                vec.push_back(st);
                return;
            }
        st.push_back(S[pos]);
        generate(vec, st, S, pos+1, len);
        st.pop_back();
        generate(vec, st, S, pos+1, len);
    }

vector<vector<int> > subsets(vector<int> &S)
    {
        vector<vector<int> > vec;
        vector<int> st;
        int len=S.size();
        sort(S.begin(),S.end(),cmp);
        generate(vec, st,S,0,len);
        return vec;
    }

File: Python/Subsets.py
#!/usr/bin/env python

import copy

class Solution:
    # @param S, a list of integer
    # @return a list of lists of integer
    def generate(self, seq, S, s, num):
        if num == len(S):
            seq.append(copy.copy(s))
            return seq
        seq = self.generate(seq, S, s, num+1)
        s.append(S[num])
        seq = self.generate(seq, S, s, num+1)
        s.pop()
        return seq
    def subsets(self, S):
        seq = []
        s = []
        S.sort()
        return self.generate(seq, S, s, 0)